File: brainfuck.c
#include <stdio.h>
#include <stdlib.h>

#define MAX_LOOP 1000
#define TAPE_LENGTH 300000

#define error(msg) exit(fprintf(stderr, "%s\n", msg))

void execute(char *program) {
    char tape[TAPE_LENGTH];
    int stack[MAX_LOOP];
    int sp = 0;
    int ip = 0;
    int dp = 0;

    int n;
    int skip = 0;

    for (;;) {
        char i = program[ip++];
        switch (i) {
            case '>':
                n = 1;
                while (program[ip] == '>') {
                    n++;
                    ip++;
                }
                if (!skip) {
                    dp += n;
                    while (dp >= TAPE_LENGTH)
                        dp -= TAPE_LENGTH;
                }
                break;

            case '<':
                n = 1;
                while (program[ip] == '<') {
                    n++;
                    ip++;
                }
                if (!skip) {
                    dp -= n;
                    while (dp < 0)
                        dp += TAPE_LENGTH;
                }
                break;

            case '+':
                n = 1;
                while (program[ip] == '+') {
                    n++;
                    ip++;
                }
                if (!skip) tape[dp] += n;
                break;

            case '-':
                n = 1;
                while (program[ip] == '-') {
                    n++;
                    ip++;
                }
                if (!skip) tape[dp] -= n;
                break;

            case ',':
                if (!skip) {
                    tape[dp] = (char)getchar();
                }
                break;

            case '.':
                if (!skip) {
                    putchar((int)tape[dp]);
                }
                break;

            case '[':
                stack[sp++] = ip;
                if (sp >= MAX_LOOP) error("too deep recursion");
                if (!tape[dp]) skip ++;
                break;

            case ']':
                if (sp <= 0) error("unmatched ']'");

                if (tape[dp]) ip = stack[sp-1];
                else sp --;
               
                if (skip) skip --;
                break;

            case 0:
                if (sp > 0) error("unmatched '[]");
                return;
        }
    }
}

int main(int argc, char** argv) {
    if (argc == 2) {
        long sz;
        char *program;
        FILE *f = fopen(argv[1], "r");
        if (!f) error("can't open file");

        fseek(f, 0, SEEK_END);
        sz = ftell(f);
        fseek(f, 0, SEEK_SET);
        program = (char*) malloc(sz+1);
        program[fread(program, 1, sz, f)] = 0;
        fclose(f);
        execute(program);
    } else {
        error("usage: brainfuck bffile");
    }

    return 0;
}

File: brainfuck.py
from array import array
from sys import stdin, stdout

class BrainfuckVM(object):
    def __init__(self):
        pass

    def execute(self, program):
        TAPE_SIZE = 30000
        data = bytearray(TAPE_SIZE)
        dp = ip = 0
        stack = array('I')
        skip = 0

        while True:
            i = program[ip]
            ip += 1
            if i == '<':
                n = 1
                while program[ip] == '<':
                    ip += 1
                    n += 1
                if skip == 0:
                    dp -= n
                    while dp < 0:
                        dp += TAPE_SIZE
            elif i == '>':
                n = 1
                while program[ip] == '>':
                    ip += 1
                    n += 1
                if skip == 0:
                    dp += n
                    dp = dp % TAPE_SIZE
            elif i == '+':
                n = 1
                while program[ip] == '+':
                    ip += 1
                    n += 1
                if skip == 0:
                    data[dp] = (data[dp] + n) % 256
            elif i == '-':
                n = 1
                while program[ip] == '-':
                    ip += 1
                    n += 1
                if skip == 0:
                    r = data[dp] - n
                    while r < 0:
                        r += 256
                    data[dp] = r
            elif i == ',':
                if skip == 0:
                    data[dp] = ord(stdin.read(1))
            elif i == '.':
                if skip == 0:
                    stdout.write(chr(data[dp]))
            elif i == '[':
                stack.append(ip)
                if data[dp] == 0:
                    skip += 1
            elif i == ']':
                if len(stack) == 0:
                    raise ValueError("no corresponding '['")
                if data[dp] == 0:
                    stack.pop()
                else:
                    ip = stack[-1]
                if skip > 0:
                    skip -= 1
            elif i == '\0':
                if len(stack) > 0:
                    raise ValueError("no corresponding ']'")
                return

if __name__ == '__main__':
    from sys import argv
    if len(argv) != 2:
        print('Usage: brainfuck bffile')
    else:
        with open(argv[1]) as bf:
            program = bf.read()

        program = program + '\0'
        vm = BrainfuckVM()
        vm.execute(program)
